src/main/java/com/datapipe/jenkins/vault/credentials/AbstractVaultTokenCredential.java

```diff
@@ -2,21 +2,69 @@
 
 import com.bettercloud.vault.Vault;
 import com.bettercloud.vault.VaultConfig;
+import com.bettercloud.vault.VaultException;
 import com.cloudbees.plugins.credentials.CredentialsScope;
 import com.cloudbees.plugins.credentials.impl.BaseStandardCredentials;
+import com.datapipe.jenkins.vault.exception.VaultPluginException;
+import edu.umd.cs.findbugs.annotations.CheckForNull;
+import hudson.Util;
+import org.kohsuke.stapler.DataBoundSetter;
 
 public abstract class AbstractVaultTokenCredential
     extends BaseStandardCredentials implements VaultCredential {
 
+    @CheckForNull
+    private String namespace;
+
     protected AbstractVaultTokenCredential(CredentialsScope scope, String id, String description) {
         super(scope, id, description);
     }
 
+    /**
+     * Get Vault namespace.
+     * @return vault namespace or null
+     */
+    @CheckForNull
+    public String getNamespace() {
+        return namespace;
+    }
+
+    /**
+     * Set namespace where auth method is mounted. If set to "/"
+     * the root namespace is explicitly forced, otherwise the
+     * namespace from the secret credential vault config is used.
+     * @param namespace vault namespace
+     */
+    @DataBoundSetter
+    public void setNamespace(String namespace) {
+        this.namespace = Util.fixEmptyAndTrim(namespace);
+    }
+
     protected abstract String getToken(Vault vault);
 
     @Override
     public Vault authorizeWithVault(VaultConfig config) {
-        Vault vault = new Vault(config);
-        return new Vault(config.token(getToken(vault)));
+        try {
+            VaultConfig authCfg = config;
+            // override vault credential namespace
+            if (namespace != null) {
+                authCfg = new VaultConfig()
+                    .address(config.getAddress())
+                    .engineVersion(config.getGlobalEngineVersion())
+                    .prefixPathDepth(config.getPrefixPathDepth())
+                    .sslConfig(config.getSslConfig())
+                    .secretsEnginePathMap(config.getSecretsEnginePathMap())
+                    .openTimeout(config.getOpenTimeout())
+                    .readTimeout(config.getReadTimeout());
+                if (!namespace.trim().equals("/")) {
+                    authCfg = authCfg.nameSpace(namespace);
+                }
+                authCfg.build();
+            }
+            String token = getToken(new Vault(authCfg));
+            return new Vault(config.token(token));
+        } catch (VaultException ve) {
+            throw new VaultPluginException("failed to authorize vault credentials", ve);
+        }
```

Comment:
Everything is fine except this part. This can be very simply written as:

```suggestion
        if (namespace != null) {
            config.nameSpace(namespace);
        }
        Vault vault = new Vault(config);
        return new Vault(config.token(getToken(vault)));
```

Reply:
I wasn't particularly proud of this, but the reason for this is for the use case where the auth credential is mounted in a different namespace than the secret we are retrieving, so we need to switch the namespace without potentially changing the namespace used by the secret read. Because the VaultConfig is shared between the auth and the secret read and the "namespace" property on the config can't be reset to null after it's already by set this could cause issues. 

The use case in our org is that the kubernetes auth is mounted on the root namespace and the secret itself is in a namespace. If the vault config was not copied then the secret read would use the the auth credential namespace (i.e. root) instead of the configured secret namespace.

It's unfortunate VaultConfig doesn't have a copy constructor or clone mechanism.

Hopefully I explained that clearly :)

Reply:
Our VaultConfiguration has a clone method and can help build the VaultConfig 🤔 Perhaps use that as vehicle to get a VaultConfig?

Reply:
I may have found a better way to handle this. It turns out `vault.auth().withNamespace(namespace)` does not effect the underlying VaultConfig settings. I refactored the code a bit to introduce a new abstract vault token class to handle this auth client configuration for credential types that need it. Let me know what you think.